### PR TITLE
feat(uat): implement subscribe publish message_is_received steps

### PR DESCRIPTION
**Pull request by bgklika**
**Issue #, if available:**
https://klika-tech.atlassian.net/browse/GGMQ-59
https://klika-tech.atlassian.net/browse/GGMQ-74
https://klika-tech.atlassian.net/browse/GGMQ-69
https://klika-tech.atlassian.net/browse/GGMQ-66

Look at https://github.com/aws-greengrass/aws-greengrass-client-device-auth/pull/259

### CODE REVIEW

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        this.connectionName = builder.connectionName;
+        this.topic = builder.topic;
+        this.topicFilter = builder.topicFilter;
+        this.content = builder.content;
+    }
+
+    /**
+     * Builder of event filter.
+     */
+    public static class Builder {
```
Comment (esmirnov-klika-tech) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:36]: Could we simply use @Builder annotation from lombok ? https://projectlombok.org/features/Builder

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    void addEvent(@NonNull Event event);
+
+    /**
+     * Gets matched events from storage immediately.
+     *
+     * @param filter the filter of events
+     * @return the array of matched events
+     */
+    Event[] getEvents(@NonNull EventFilter filter);
```
Comment (esmirnov-klika-tech) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java:36]: I'd like to change arrays to something from Collections API. It is a common practice to use List - https://docs.oracle.com/javase/8/docs/api/java/util/List.html

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+     * Waiting for matched events.
+     *
+     * @param filter the filter of events
+     * @param timeout the maximum time to wait
+     * @param unit the time unit of the timeout argument
+     * @return the array of matched events
+     * @throws TimeoutException if the wait timed out
+     * @throws InterruptedException if the current thread was interrupted while waiting
+     */
+    Event[] awatingEvents(@NonNull EventFilter filter, long timeout, @NonNull TimeUnit unit)
```
Comment (esmirnov-klika-tech) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java:48]: It is nice to have, but for tests we can use just sleep with following getEvents call

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Filter of events in event storage.
+ */
+@Getter
+public final class EventFilter {
+    private final Event.Type type;
+    private final String connectionName;
+    private final String topic;
+    private final String topicFilter;
+    private final byte[] content;
```
Comment (esmirnov-klika-tech) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:22]: Can we use String for content?

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
```
Comment (esmirnov-klika-tech) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:23]: Could you add timestampFrom and timestampTo please?

**Review by esmirnov-klika-tech — APPROVED** [review]

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java (hunk 1 above)
Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:36]: lombok Build is not sutiable in that case due to we have overloaded withContent() method with String and byte[] arguments. I checked lombok manual and did not found ability to add extra methods to it builder.

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java (hunk 4 above)
Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:22]: By nature it is array of bytes not a string. For example sparkplug messages can be sent over MQTT.

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java (hunk 3 above)
Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java:48]: It really required due to you can't predict when message will be received.
Sleep will suspend test execution all the time, passing all tests will takes so long time.

Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java:48]: For example: you waiting for 60 seconds max. With sleep(60) is takes always at least 60 seconds.
With awatingEvents(60) it takes only 10 seconds if message will be received after 10 seconds.
We save 50 seconds on one test. With greengrass we can have thousands of tests.

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java (hunk 2 above)
Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventStorage.java:36]: Fixed in a5267e0af6fd266ee4f0465106cf4f315bfbf526

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java (hunk 5 above)
Comment (bgklika) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/EventFilter.java:23]: Added in 77e7619407a9e819ebb329ac24a1e28bd4e5244d

**Review by esmirnov-klika-tech — APPROVED** [review]

**Review by auarbekov — APPROVED** [review]

**Review by esmirnov-klika-tech — APPROVED** [review]

**Review by esmirnov-klika-tech — APPROVED** [review]

**Review by esmirnov-klika-tech — APPROVED** [review]

**Review by auarbekov — DISMISSED** [review]

**Review by auarbekov — DISMISSED** [review]

File: uat/README.md [hunk 6]
```diff
@@ -20,31 +31,33 @@ $Env:AWS_ACCESS_KEY_ID=<AWS_ACCESS_KEY_ID>
...
 ```
 
-Execute the UATs by running the following command from the root of the project.
+### Run scenarios
+Execute the UATs by running the following commands from the root of the project.
 
 ```
-java -Dggc.archive=<path-to-nucleus-zip> -Dtest.log.path=<path-to-test-results-folder> -Dtags=Auth -jar <path-to-test-jar>
-java -Dggc.archive=./greengrass-nucleus-latest.zip -Dtest.log.path=./logs -Dtags=Auth -jar uat/testing-features/target/client-devices-auth-testing-features.jar
+java -Dggc.archive=<path-to-nucleus-zip> -Dtest.log.path=<path-to-test-results-folder> -Dtags=GGMQ -jar <path-to-test-jar>
```
[the remark this answers is not in the document]
Comment (mzyl13) [reply uat/README.md:55]: Now this is up-to-date command template

**Pull request closed without merging**